pkg/discovery/systemd.go

```diff
@@ -50,111 +45,65 @@ func (c *SystemdConfig) NewDiscoverer(d DiscovererOptions) (Discoverer, error) {
 }
 
 func (c *SystemdDiscoverer) Run(ctx context.Context, up chan<- []*Group) error {
-	ticker := time.NewTicker(time.Second)
-	defer ticker.Stop()
-
-	for {
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		case <-ticker.C:
-		}
-
-		units, err := c.units(ctx)
-		if err != nil {
-			return fmt.Errorf("failed to list units: %w", err)
-		}
-		newSourceList := map[string]struct{}{}
-		var targetGroups []*Group
-		for _, unit := range units {
-			labelset := model.LabelSet{
-				"systemd_unit": model.LabelValue(unit),
-			}
-			pids, err := c.pids(ctx, unit)
-			if err != nil {
-				level.Debug(c.logger).Log("msg", "failed to get PIDs for unit", "unit", unit, "err", err)
-				continue
-			}
-
-			newSourceList[unit] = struct{}{}
-
-			targetGroups = append(targetGroups, &Group{
-				Targets: []model.LabelSet{{}},
-				Labels:  labelset,
-				Source:  unit,
-				PIDs:    pids,
-			})
-		}
-
-		// Add empty groups for targets which have been removed since the previous run.
-		for unit := range c.oldSourceList {
-			if _, ok := newSourceList[unit]; !ok {
-				targetGroups = append(targetGroups, &Group{Source: unit})
-			}
-		}
-		c.oldSourceList = newSourceList
-
-		up <- targetGroups
-	}
-}
-
-func (c *SystemdDiscoverer) units(ctx context.Context) ([]string, error) {
-	ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
-	defer cancel()
-
-	cmd := exec.CommandContext(ctx, "systemctl", "list-units", "--type=service", "--state=running")
-	stdout, err := cmd.StdoutPipe()
+	conn, err := systemd.NewWithContext(ctx)
 	if err != nil {
-		return nil, err
-	}
-	if err := cmd.Start(); err != nil {
-		return nil, err
+		level.Error(c.logger).Log("msg", "failed to connect to systemd D-Bus API", "err", err)
+		return err
```

Comment:
```suggestion
		return fmt.Errorf("failed to connect to systemd D-Bus API, %w", err)
```
Errors should be handled once and logging is error handling, so let's propagate the error to the caller and let it handle the error.